feature/SignalR

the-man-w-laughs opened the pull request:
SignalR is used to implement real-time chatting within tasks.

the-man-w-laughs committed:
Add base for SignalR implementation.

Major changes:
- add SignalR registration;
- add hub codebase.

the-man-w-laughs committed:
Refactor code

Major changes:
- add constructors to some models;
- add CreatedAt initialization inside automapper profiles or inside constructors.

the-man-w-laughs committed:
Refactor TeamTaskHub

Major changes:
- update models and Dtos fields names;
- add some new Dtos;
- add services to handle simple operations like find and check for null (to unload command handlers);
- make mediatr command handlers return Dtos instead of ids;
- add comments to mediatr handlers.

the-man-w-laughs committed:
Final implementation of SignalR

Major changes:
- add connection handler;
- add methods to create, delete, update comments.

the-man-w-laughs committed:
Add web socket configuration to api gateway

Major changes:
- change ocelot.json configuration;
- adjust CommentsHub.

the-man-w-laughs committed:
Refactor ProjectReportInfo

the-man-w-laughs committed:
Completely remove business logic from repositories.

the-man-w-laughs committed:
Remove unnecessary using statements from ReportHub.

mebry commented on src/services/ReportHub/ReportHub.DAL/Models/ProjectReportInfo.cs:
```diff
@@ -4,6 +4,15 @@ namespace ReportHub.DAL.Models;
+
     public int ProjectId { get; set; }
     public int ProjectAuthorId { get; set; }
     public DateTime CreatedAt { get; set; }
```
When describing a class, use the following structure:
· Declare all class members, including readonly or constants
· Declare all public properties, including auto-properties
· Declare all constructor(s)
· Declare the destructor, if any
· Declare all public methods
· Declare all protected methods
· Declare all private methods

mebry commented on src/services/ReportHub/ReportHub.DAL/Models/Report.cs:
```diff
@@ -2,6 +2,12 @@ namespace ReportHub.DAL.Models;
+        GeneratedAt = DateTime.Now;
+    }
+
     public string Path { get; set; }
```
The same

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/ICommentsHub.cs:
```diff
@@ -0,0 +1,13 @@
+        Task Connection(TaskModelResponseDto task);
+        Task CreateComment(CommentResponseDto commentResponseDto);
+        Task UpdateComment(CommentResponseDto commentResponseDto);
+        Task DeleteComment(CommentResponseDto commentResponseDto);
```
Add the suffix "Async" to the names of methods

mebry commented on src/services/TeamHub/TeamHub.BLL/Extensions/HubContextExtensions.cs:
```diff
@@ -0,0 +1,13 @@
+
+        public static int GetTaskId(this IDictionary<object, object?> items) =>
+            Convert.ToInt32(items[HubContextItems.TaskId]);
+    }
```
It's better to use curly braces to adhere to a conventional coding style

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/ITaskHandlerService.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace TeamHub.BLL.Contracts
+{
+    public interface ITaskHandlerService
+    {
+        Task<TaskHandler> AddTaskHandlerAsync(
+            TeamMember teamMember,
+            int taskId,
+            CancellationToken cancellationToken
+        );
+        Task<TaskHandler> RemoveTaskHandlerAsync(
+            TeamMember teamMember,
+            int taskId,
+            CancellationToken cancellationToken
+        );
```
You don't need to create your own services for this. Describe the implementation of creation and deletion in the handler for your commands

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/ICommentService.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace TeamHub.BLL.Contracts
+{
+    public interface ICommentService
+    {
+        Task<Comment> GetCommentAsync(int commentId, CancellationToken cancellationToken);
+    }
```
Rename the method in the service. This service is used to check for existence and return an object. You can rename it to: GetCommentAndCheckIfItExistsAsync.
Additionally, it's better to change the service name to make it explicitly clear what its purpose will be

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/IProjectService.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace TeamHub.BLL.Contracts
+{
+    public interface IProjectService
+    {
+        Task<Project> GetProjectAsync(int projectId, CancellationToken cancellationToken = default);
+    }
```
Similarly to the ICommentService

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/ITaskService.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace TeamHub.BLL.Contracts
+{
+    public interface ITaskService
+    {
+        Task<TaskModel> GetTaskAsync(int taskId, CancellationToken cancellationToken);
+    }
```
Similarly to the ICommentService

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/ITeamMemberService.cs:
```diff
@@ -0,0 +1,25 @@
+            int userId,
+            int projectId,
+            CancellationToken cancellationToken
+        );
```
You don't need to create your own services for this. Describe the implementation of creation and deletion in the handler for your commands

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/IUserService.cs:
```diff
@@ -0,0 +1,9 @@
+    public interface IUserService
+    {
+        Task<User> GetUserAsync(int userId, CancellationToken cancellationToken);
+    }
```
Similarly to the ICommentService

the-man-w-laughs committed:
Implement Mentor's Recommendations

Major changes:
- update class declarations to adhere convension;
- change lambda to curly braces to adhere code convension;
- set more appropriate service names and theirs methods;
- add async to asynchronous methods;
- remove service methods that are used only once and put the implementation inside commnd handlers.

the-man-w-laughs committed:
Merge branch 'feature/hangfire' into feature/SignalR
